# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: parser/html/nsHtml5TreeOpExecutor.h
# Commit: 9a8abd87cc79
# Full Hash: 9a8abd87cc7935f29b94248c1a6f8203faa14403
# Author: Henri Sivonen <hsivonen@hsivonen.fi>
# Date: 2021-12-08 21:43:06
# Regressor Bug: 1701828
# File Overlap Count: 1
# Description:
#   Bug 1701828 - meta charset rewrite. r=smaug
#   
#   Implements https://github.com/whatwg/html/issues/6962 . Improves performance
#   when <meta charset> occurs in head but after the first kilobyte and aligns
#   behavior better with WebKit and Blink.
# ==============================================================================

diff --git a/parser/html/nsHtml5TreeOpExecutor.h b/parser/html/nsHtml5TreeOpExecutor.h
--- a/parser/html/nsHtml5TreeOpExecutor.h
+++ b/parser/html/nsHtml5TreeOpExecutor.h
@@ -55,16 +55,31 @@ class nsHtml5TreeOpExecutor final
   static uint32_t sTimesFlushLoopInterrupted;
 #endif
 
   /**
    * Whether EOF needs to be suppressed
    */
   bool mSuppressEOF;
 
+  /**
+   * Set to true if CommitToInternalEncoding() was unable to flush all
+   * pending operations. In that case, this flag signals to
+   * ContinueInterruptedParsingAsync() that another attempt to call
+   * CommitToInternalEncoding() (from the event loop) is need.
+   *
+   * This may happen in particular when extensions inject scripts
+   * to the document, and the script injection prevents
+   * CommitToInternalEncoding() from completing in one attempt.
+   *
+   * This can also happen as a result of the flush being slow enough
+   * that the flush is interrupted based on time.
+   */
+  bool mPendingEncodingCommitment;
+
   bool mReadingFromStage;
   nsTArray<nsHtml5TreeOperation> mOpQueue;
   nsHtml5StreamParser* mStreamParser;
 
   /**
    * URLs already preloaded/preloading.
    */
   nsTHashSet<nsCString> mPreloadedURLs;
@@ -174,29 +189,34 @@ class nsHtml5TreeOpExecutor final
   void StartLayout(bool* aInterrupted);
 
   void PauseDocUpdate(bool* aInterrupted);
 
   void FlushSpeculativeLoads();
 
   void RunFlushLoop();
 
+  void RunFlushLoopOrCommitToInternalEncoding();
+
   nsresult FlushDocumentWrite();
 
+  void CommitToInternalEncoding();
+
   void MaybeSuspend();
 
   void Start();
 
   void NeedsCharsetSwitchTo(NotNull<const Encoding*> aEncoding, int32_t aSource,
                             uint32_t aLineNumber);
 
   void MaybeComplainAboutCharset(const char* aMsgId, bool aError,
                                  uint32_t aLineNumber);
 
-  void ComplainAboutBogusProtocolCharset(mozilla::dom::Document*);
+  void ComplainAboutBogusProtocolCharset(mozilla::dom::Document* aDoc,
+                                         bool aUnrecognized);
 
   void MaybeComplainAboutDeepTree(uint32_t aLineNumber);
 
   bool HasStarted() { return mStarted; }
 
   bool IsFlushing() { return mFlushState >= eInFlush; }
 
 #ifdef DEBUG